packages/adapter/src/lib/_Types.ts

```diff
@@ -46,6 +50,65 @@ export interface AdapterOptions {
     error?: ioBroker.ErrorHandler;
 }
 
+type MessageUnsubscribeReason = 'client' | 'disconnect';
+export type ClientUnsubscribeReason = MessageUnsubscribeReason | 'clientSubscribeError';
+type UserInterfaceClientUnsubscribeReason = ClientUnsubscribeReason | 'timeout';
+
+export interface UserInterfaceSubscribeInfo {
+    /** The client id, which can be used to send information to clients */
+    clientId: string;
+    /** The message used for subscription */
+    message: ioBroker.Message;
+}
+
+export type UserInterfaceClientSubscribeHandler = (
+    subscribeInfo: UserInterfaceSubscribeInfo
+) => UserInterfaceClientSubscribeReturnType | Promise<UserInterfaceClientSubscribeReturnType>;
+
+export interface UserInterfaceClientSubscribeReturnType {
+    /** If the adapter has accepted the client subscription */
+    accepted: boolean;
+    /** Optional heartbeat, if set, the client needs to re-subscribe every heartbeat interval */
+    heartbeat?: number;
+}
+
+type UserInterfaceUnsubscribeInfoBaseObject = {
+    /** The handler id, which can be used to send information to clients */
+    clientId: string;
+};
+
+export type UserInterfaceUnsubscribeInfo = UserInterfaceUnsubscribeInfoBaseObject &
+    (
+        | {
+              /** Reason for unsubscribe */
+              reason: Exclude<UserInterfaceClientUnsubscribeReason, ClientUnsubscribeReason>;
```

Comment:
```suggestion
              reason: Exclude<UserInterfaceClientUnsubscribeReason, ClientUnsubscribeReason>;
              message?: undefined;
```
avoids having to test `message in something` before accessing `message`